design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,256 @@
+3. A developer runs a simple command to run the tests, this results in either success or failure, in both cases the outcome is reported and any resources used in the test are cleaned up (unless requested to stay). In the case of failure breadcrumbs are provided fo where in the test the failure occurred.
+4. A developer adds the test to CI and pushes the branch got Github
+5. A repetition of 3 occurs except this case it in CI
+
+## Walkthroughs
+
+TBD
+
+# Technical Design
+
+## Architecture
+
+[ Service ] is any software that tests rely on e.g. Conjur, Postgres etc.
+
+[ Service Config ] specifies details about a service e.g. for Conjur `CONJUR_ACCOUNT`, `version` etc. Conjur will have different configuration options to Postgres. These options will need to be documented. It's important for users of the Service Engine to know what they're able to configure on a service. Config also specifies the lifecyle method of the interface (start, stop, clean)
+
+[ Test Runner ] is software that provides a conventional way for defining and running test cases. They typically come with capabilities to measure test metrics, test outcomes and code coverage. This will also allow historical tracking. An example of a test-runner is Cucumber. Test cases can be defined in Gherkin
+
+[ Service Engine ] uses [ Service Config ] to create an instance of a [ Service ] 
+
+[ Test Case ] specifies a test case and contains some annotations detailing required [ Service ]s 
+
+[ Test Runner ] gathers annotations from [ Service ]s from [ Test Case ]s and issues requests for services to the [ Service Engine ] before running all [ Test Case ].
+
+[ Service Engine ] uses [ Service Config ] to create an instance of a [ Service ].
+
+Here is an example of the components in this architecture working together.
+0. A service definition exists for conjur and mysql
+```bash
+# Conjur service definition.  This is supplied to the start_services function and defines the service to be started
+declare -A conjur_service_config
+conjur_service_config=(
+["service_name"]="conjur_abc",
+["config"]="conjur.yaml",
+["start"]="conjur_start.sh",
+["stop"]="conjur_stop.sh"
+)
+
+declare -A mysql_service_config
+mysql_service_config=(
+["service_name"]="mysql_abc",
+["config"]="mysql.yaml",
+["start"]="mysql_start.sh",
+["stop"]="mysql_stop.sh"
+)
+``` 
+1. There is a test case written in Gherkin for validating using the MySQL service connector with Conjur.
+```
+Feature: Guess the word
+  
+  @conjur @mysql @mysql-conjur
+  Scenario: Connection to MySQL
+    Given Secretless is running as a sidecar in Kubernetes
+    Then the app is able to connect to MYSQL on "localhost:332"
+
+```
+2. Cucumber will register that there's a test case that requires Conjur, MySQL and MySQL credentials in Conjur.
+3. The developer calls `cucumber` to start the testss
+4. As a result of the annotations Cucumber will first call the engine to create the services.
+5. The configuration values
+ 
```

Comment:
this example is amazing, and does a lot to help clarify how it will work in practice.

is there a way to link the tags in the example (`@conjur @mysql @mysql-conjur`) to the service definitions declared above? it's not immediately clear how they're related or how the link is made

Reply:
Updated